pkg/ip/ip.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2016-2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ip
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"sort"
+
+	"github.com/asaskevich/govalidator"
+)
+
+const (
+	ipv4BitLen = 8 * net.IPv4len
+	ipv6BitLen = 8 * net.IPv6len
+)
+
+// ByMask is used to sort a list of IP networks by the size of their masks.
+type ByMask []*net.IPNet
+
+func (s ByMask) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+func (s ByMask) Less(i, j int) bool {
+	iPrefixSize, _ := s[i].Mask.Size()
+	jPrefixSize, _ := s[j].Mask.Size()
+	if iPrefixSize == jPrefixSize {
+		byteArrComp := bytes.Compare(s[i].IP, s[j].IP)
+		if byteArrComp < 0 {
+			return true
+		}
+		return false
+	}
+	return iPrefixSize < jPrefixSize
+}
+
+func (s ByMask) Len() int {
+	return len(s)
+}
+
+// RemoveCIDRs removes the specified CIDRs from another set of CIDRs. If a CIDR to remove is not
+// contained within the CIDR, the CIDR to remove is ignored. A slice of CIDRs is
+// returned which contains the set of CIDRs provided minus the set of CIDRs which
+// were removed.
+func RemoveCIDRs(allowCIDRs, removeCIDRs []*net.IPNet) ([]*net.IPNet, error) {
+
+	// Ensure that we iterate through the provided CIDRs in order of largest
+	// subnet first.
+	sort.Sort(ByMask(removeCIDRs))
+
+PreLoop:
+	// Remove CIDRs which are contained within CIDRs that we want to remove;
+	// such CIDRs are redundant.
+	for j, removeCIDR := range removeCIDRs {
+		for i, removeCIDR2 := range removeCIDRs {
+			if i == j {
+				continue
+			}
+			if removeCIDR.Contains(removeCIDR2.IP) {
+				removeCIDRs = append(removeCIDRs[:i], removeCIDRs[i+1:]...)
+				// Re-trigger loop since we have modified the slice we are iterating over.
+				goto PreLoop
+			}
+		}
+	}
+
+Loop:
+	for j, remove := range removeCIDRs {
+		for i, allowCIDR := range allowCIDRs {
+
+			// Don't allow comparison of different address spaces.
+			if !govalidator.IsIPv4(allowCIDR.IP.String()) && govalidator.IsIPv4(remove.IP.String()) ||
```

Comment:
This can be done in `Validate()` so you don't need to error out here.

Reply:
This is in its own package, though. If we want to use these functions elsewhere for whatever reason, it is useful to have an error returned.